File: VarianceExamples.java
import java.util.ArrayList;
import java.util.List;

/**
 * Variance in Java: covariant arrays, invariant generics, wildcards, PECS.
 */
public class VarianceExamples {

    static class Animal {
        String name() { return "Animal"; }
    }

    static class Cat extends Animal {
        @Override String name() { return "Cat"; }
    }

    static class Dog extends Animal {
        @Override String name() { return "Dog"; }
    }

    // --- Covariant arrays (Java's mistake) ---

    static void covariantArraysProblem() {
        Animal[] animals = new Cat[3];      // compiles — Cat[] IS-A Animal[]
        animals[0] = new Cat();             // fine
        // animals[1] = new Dog();          // 💥 ArrayStoreException at RUNTIME
    }

    // --- Invariant generics (the fix) ---

    static void invariantGenerics() {
        List<Cat> cats = new ArrayList<>();
        cats.add(new Cat());
        // List<Animal> animals = cats;     // ❌ compile error — invariant
    }

    // --- ? extends T — covariant / producer / read-only ---

    static void printAll(List<? extends Animal> animals) {
        for (Animal a : animals) {
            System.out.println(a.name());   // ✅ reading is safe
        }
        // animals.add(new Cat());          // ❌ writing is blocked
    }

    // --- ? super T — contravariant / consumer / write-only ---

    static void addCats(List<? super Cat> list) {
        list.add(new Cat());                // ✅ writing Cats is safe
        // Cat c = list.get(0);             // ❌ reading as Cat is blocked
        Object o = list.get(0);             // ✅ only guaranteed Object
    }

    // --- Plain T — invariant / read + write ---

    static void swap(List<Animal> list, int i, int j) {
        Animal temp = list.get(i);
        list.set(i, list.get(j));
        list.set(j, temp);
    }

    public static void main(String[] args) {
        // Producer Extends
        System.out.println("=== ? extends (covariant) ===");
        List<Cat> cats = List.of(new Cat(), new Cat());
        printAll(cats);

        // Consumer Super
        System.out.println("\n=== ? super (contravariant) ===");
        List<Animal> animals = new ArrayList<>();
        addCats(animals);
        System.out.println("Size after addCats: " + animals.size());

        // Invariant
        System.out.println("\n=== invariant (read + write) ===");
        List<Animal> mixed = new ArrayList<>(List.of(new Cat(), new Dog()));
        swap(mixed, 0, 1);
        System.out.println("After swap: " + mixed.get(0).name() + ", " + mixed.get(1).name());
    }
}

File: VarianceExamples.kt
/**
 * Variance in Kotlin: out (covariant), in (contravariant), invariant default.
 */

open class Animal { open fun name() = "Animal" }
class Cat : Animal() { override fun name() = "Cat" }
class Dog : Animal() { override fun name() = "Dog" }

// --- out = covariant = "I only produce T" ---

class Cage<out T>(private val animal: T) {
    fun look(): T = animal          // ✅ T in output position
    // fun put(a: T) { }            // ❌ compile error — T can't go in
}

// --- in = contravariant = "I only consume T" ---

interface Handler<in T> {
    fun handle(t: T)                // ✅ T in input position
    // fun produce(): T             // ❌ compile error — T can't come out
}

// --- no modifier = invariant = "I do both" ---
// MutableList<T> is invariant — can't assign MutableList<Cat> to MutableList<Animal>

// --- Declaration-site vs use-site ---

interface Source<out T> {
    fun produce(): T
}

// Use-site projection on an invariant type
fun copy(from: MutableList<out Animal>, to: MutableList<in Animal>) {
    for (a in from) to.add(a)
}

fun main() {
    // Covariance: Cage<Cat> IS-A Cage<Animal>
    println("=== out (covariant) ===")
    val catCage: Cage<Cat> = Cage(Cat())
    val animalCage: Cage<Animal> = catCage  // ✅
    println(animalCage.look().name())

    // Contravariance: Handler<Animal> IS-A Handler<Cat>
    println("\n=== in (contravariant) ===")
    val animalHandler = object : Handler<Animal> {
        override fun handle(t: Animal) {
            println("Handling ${t.name()}")
        }
    }
    val catHandler: Handler<Cat> = animalHandler  // ✅ reversed
    catHandler.handle(Cat())

    // Invariance: MutableList<Cat> is NOT MutableList<Animal>
    println("\n=== invariant ===")
    val cats: MutableList<Cat> = mutableListOf(Cat())
    // val animals: MutableList<Animal> = cats  // ❌ won't compile

    // Covariant read-only list
    println("\n=== List<out E> stdlib ===")
    val catList: List<Cat> = listOf(Cat(), Cat())
    val animalList: List<Animal> = catList   // ✅ List is covariant
    println("Animals: ${animalList.size}")
}
